Use Roslyn support for RuntimeHelpers.CreateSpan

stephentoub opened the pull request:
As of https://github.com/dotnet/roslyn/pull/61414, the optimization that previously extended only to byte/sbyte/bool will now also support char/ushort/short/uint/int/ulong/long/float/double.  This PR uses it throughout dotnet/runtime.  However, because of endianness issues, this requires the RuntimeHelpers.CreateSpan method that only exists as of .NET 7, so trying to use this optimization in builds targeting < .NET 7 will be a significant deoptimization.  Until we have dedicated syntax (ala https://github.com/dotnet/csharplang/issues/5295) and/or an analyzer (https://github.com/dotnet/runtime/issues/69577) to avoid such issues, we need to be very careful where this applied.

I'm putting this up as a draft now for feedback, but we must not merge this until we've picked up a Roslyn compiler that has the aforementioned improvement.

Closes https://github.com/dotnet/runtime/issues/60948
cc: @VSadov, @jcouv, @davidwrighton, @GrabYourPitchforks

GrabYourPitchforks commented on src/libraries/Common/src/System/Drawing/KnownColorTable.cs:
```diff
@@ -466,11 +466,12 @@ internal static class KnownColorTable
         {
             Debug.Assert((argb & Color.ARGBAlphaMask) == Color.ARGBAlphaMask);
-            Debug.Assert(s_colorValueTable.Length == ColorKindTable.Length);
+            Debug.Assert(ColorValueTable.Length == ColorKindTable.Length);
```
(No action required on your part - just dropping this note as a general reminder.)

We'll want to confirm that `ColorValueTable.Length` is replaced with a literal int burned into the codegen. I remember the JIT needing to jump through some hoops to optimize the `((ROS<byte>)new byte[] { /* ... */ }).Length` case, and I suspect it was complex enough that we might not get it for free here.

stephentoub replied:
This:
```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

[DisassemblyDiagnoser(maxDepth: 3)]
public class Program
{
    static void Main(string[] args) => BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly).Run(args);

    private static ReadOnlySpan<char> Values => new char[] { 'a', 'b', 'c', 'd', 'e' };

    [Benchmark]
    public int GetLength() => Values.Length;
}
```
results in this IL for `Values`:
```
	IL_0000: ldtoken field valuetype '<PrivateImplementationDetails>'/'__StaticArrayInitTypeSize=10' '<PrivateImplementationDetails>'::DC0F42A41F058686A364AF5B6BD49175C5B2CF3C4D5AE95417448BE3517B4008
	IL_0005: call valuetype [System.Runtime]System.ReadOnlySpan`1<!!0> [System.Runtime]System.Runtime.CompilerServices.RuntimeHelpers::CreateSpan<char>(valuetype [System.Runtime]System.RuntimeFieldHandle)
	IL_000a: ret
```
and this asm for GetLength:
```
## .NET 7.0.0 (42.42.42.42424), X64 RyuJIT
```assembly
; Program.GetLength()
       mov       eax,5
       ret
; Total bytes of code 6
```
So at least on this small repro, it looks good.

GrabYourPitchforks replied:
That's awesome! Thanks for confirming.

GrabYourPitchforks commented on src/libraries/System.Data.Common/src/System/Data/SQLTypes/SQLGuid.cs:
```diff
@@ -125,16 +121,22 @@ public static SqlGuid Parse(string s)
         private static EComparison Compare(SqlGuid x, SqlGuid y)
         {
+            // Comparison orders.
+            ReadOnlySpan<int> rgiGuidOrder = new int[16] { 10, 11, 12, 13, 14, 15, 8, 9, 6, 7, 4, 5, 0, 1, 2, 3 };
```
Nit: We could make this a `ROS<byte>` and check it in now, and we wouldn't have to worry about downlevel weirdness or compiler support.

huoyaoyuan commented:
Current implementation throws hardly for big endian. It preserves the option to copy the field and reverse the endianness in memory, right? Just make sure this feature is not impossible or extremely inefficient to implement for big endian.

stephentoub commented:
> Current implementation throws hardly for big endian. It still allows the option to copy the field and reverse the endianness in memory, right? Just make sure this feature is not impossible or extremely inefficient to implement for big endian.

The whole purpose of the CreateSpan API is to address endianness. coreclr doesn't support big endian at all today (well beyond this particular method), but mono does, including with this method.
https://github.com/dotnet/runtime/blob/5adac4ac39c7473d3b17ab8875aeeaae79d27187/src/mono/mono/metadata/icall.c#L960

GrabYourPitchforks commented on src/libraries/System.Private.CoreLib/src/System/Globalization/EastAsianLunisolarCalendar.cs:
```diff
@@ -11,9 +11,8 @@ public abstract class EastAsianLunisolarCalendar : Calendar
-        private static readonly int[] s_daysToMonth365 = { 0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334 };
-
-        private static readonly int[] s_daysToMonth366 = { 0, 31, 60, 91, 121, 152, 182, 213, 244, 274, 305, 335 };
+        private static ReadOnlySpan<int> DaysToMonth365 => new int[] { 0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334 };
```
There are 6 copies of this pattern throughout corelib. Might be a good opportunity for consolidation into a utility class.

GrabYourPitchforks commented on src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameParser.cs:
```diff
@@ -90,7 +89,7 @@ private AssemblyNameParts Parse()
                 ThrowInvalidAssemblyName();
 
-            if (name == string.Empty || name.IndexOfAny(s_illegalCharactersInSimpleName) != -1)
+            if (name == string.Empty || name.AsSpan().IndexOfAny('/', '\\', ':') != -1)
```
Nit: changing the first part to `string.IsNullOrEmpty(name)` would offer additional defense against somebody accidentally violating the nullability annotation. The previous code would've thrown if _name_ somehow ended up null, but the new code will not.

GrabYourPitchforks commented on src/libraries/System.Text.RegularExpressions/tools/DataTable.cs:
```diff
@@ -110,8 +110,15 @@ void EmitFileHeadersAndUsings(StreamWriter writer)
-                writer.Write("        private static ushort[] EquivalenceFirstLevelLookup { get; } = new ushort[" + firstLevelLookupTable.Count + "]\n        {\n");
+                List<ushort> firstLevelLookupTable = FlattenFirstLevelLookupTable();
+
+                writer.Write($"#if REGEXGENERATOR\");
```
Looks like some stray slashes at the end of the quotes here.

stephentoub committed:
Use Roslyn support for RuntimeHelpers.CreateSpan

The optimization that previously extended only to byte/sbyte/bool now also supports char/ushort/short/uint/int/ulong/long/float/double.

marek-safar commented on src/libraries/System.Net.HttpListener/src/System/Net/HttpListenerResponse.cs:
```diff
@@ -79,20 +79,11 @@ public bool SendChunked
-            return true;
-        }
+
+        private static bool CanSendResponseBody(int responseCode) =>
```
It might be worth extracting changes likes this one from this PR

stephentoub replied:
Good suggestion.  Done in https://github.com/dotnet/runtime/pull/72727.

marek-safar commented on src/libraries/System.Private.CoreLib/src/System/Globalization/DateTimeParse.cs:
```diff
@@ -5455,12 +5453,14 @@ internal bool MatchSpecifiedWords(string target, bool checkWordBoundary, ref int
                 // Check word by word
                 int targetPosition = 0;                 // Where we are in the target string
                 int thisPosition = Index;         // Where we are in this string
-                int wsIndex = target.IndexOfAny(WhiteSpaceChecks, targetPosition);
```
Wondering if there is analyzer for such patterns

stephentoub commented:
It looks like the compiler dependency won't land in time for the compiler change to be merged. Closing for .NET 7.
